### PR TITLE
Make sure ctx in commands are derived from req.Context

### LINKED ISSUE #505: Add a context to cmds.Request
Currently (unless I'm missing something) we have no way of allowing the user to cancel a command. If we add in a context to the request handler it will allow us to have a mechanism to cancel commands with.

### PR DESCRIPTION
**Pull request by rht**
Fixes #505

### CODE REVIEW

**Comment by rht** [comment]
@whyrusleeping there are still several left, pending. And how to add a context to https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/metadata_test.go#L52?

**Comment by rht** [comment]
Or keep the option that ipfsnode can be initialized without ctx. and modify context() method to assign context.TODO() when ctx is nil

File: unixfs/io/dirbuilder.go [hunk 1]
```diff
@@ -29,22 +27,13 @@ func NewDirectory(dserv mdag.DAGService) *directoryBuilder {
 }
 
 // AddChild adds a (name, key)-pair to the root node.
-func (d *directoryBuilder) AddChild(name string, k key.Key) error {
-	// TODO(cryptix): consolidate context managment
-	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
```
Comment (rht) [inline unixfs/io/dirbuilder.go]: no sure if `WithTimeout` should be removed.
In other places sometimes it is 30 seconds or a minute, or no `WithTimeout` at all.

Comment (whyrusleeping) [inline unixfs/io/dirbuilder.go]: Ideally, most of the places we do this (context.TODO with a timeout) should be replaced with a passed in context. and the caller can decide the timeout if they want.

**Comment by whyrusleeping** [comment]
One thing to note, we should try to avoid node.Context() if we can, as its the 'global' context. We should prefer a finer grained context when we can get it.

**Comment by rht** [comment]
though for now, node.Context() is more fine grained than context.tODO()

**Comment by whyrusleeping** [comment]
for sure! just wanted to make a note of that.

File: unixfs/mod/dagmodifier.go [hunk 2]
```diff
@@ -460,7 +460,7 @@ func dagTruncate(nd *mdag.Node, size uint64, ds mdag.DAGService) (*mdag.Node, er
 	var modified *mdag.Node
 	ndata := new(ft.FSNode)
 	for i, lnk := range nd.Links {
-		ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
+		_ctx, cancel := context.WithTimeout(ctx, time.Minute)
```
Comment (jbenet) [inline unixfs/mod/dagmodifier.go]: maybe this timeout one is one to remove? 

We'd still want a context derived from the one passed in, so things can be cancelled when this func returns

File: unixfs/io/dirbuilder.go (hunk 1 above)
Comment (jbenet) [inline unixfs/io/dirbuilder.go]: :+1: to removing this timeout.

File: core/coreunix/metadata.go [hunk 3]
```diff
@@ -44,7 +44,7 @@ func AddMetadataTo(n *core.IpfsNode, skey string, m *ft.Metadata) (string, error
 func Metadata(n *core.IpfsNode, skey string) (*ft.Metadata, error) {
 	ukey := key.B58KeyDecode(skey)
 
-	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
```
Comment (jbenet) [inline core/coreunix/metadata.go]: i think this timeout can also be removed

File: core/coreunix/metadata.go [hunk 4]
```diff
@@ -14,7 +14,7 @@ import (
 func AddMetadataTo(n *core.IpfsNode, skey string, m *ft.Metadata) (string, error) {
 	ukey := key.B58KeyDecode(skey)
 
-	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
```
Comment (jbenet) [inline core/coreunix/metadata.go]: i think this timeout can also be removed

Comment (jbenet) [inline core/coreunix/metadata.go]: replace with just a func-scoped ctx:

```go
ctx, cancel := context.WithCancel(n.Context())
defer cancel()
```

File: core/coreunix/add.go [hunk 5]
```diff
@@ -95,30 +94,21 @@ func AddWrapped(n *core.IpfsNode, r io.Reader, filename string) (string, *merkle
...
-	return node, nil
 }
 
 func addNode(n *core.IpfsNode, node *merkledag.Node) error {
-	err := n.DAG.AddRecursive(node) // add the file to the graph + local storage
-	if err != nil {
+	if err := n.DAG.AddRecursive(node); err != nil { // add the file to the graph + local storage
 		return err
 	}
-	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
```
Comment (jbenet) [inline core/coreunix/add.go]: this should not be a timeout, but rather a context derived from the request.

**Comment by jbenet** [comment]
thanks for this PR, much needed cleanup

File: core/coreunix/metadata.go (hunk 4 above)
Comment (rht) [inline core/coreunix/metadata.go]: Notice that this newly prepared context *always* happens before `DAG.Get`. To clean this up is to put it inside `DAG.Get`

File: core/coreunix/metadata.go (hunk 3 above)
Comment (rht) [inline core/coreunix/metadata.go]: same as above, because `DAG.Get`

File: unixfs/mod/dagmodifier.go (hunk 2 above)
Comment (rht) [inline unixfs/mod/dagmodifier.go]: `GetNode` basically calls `DAG.Get`, so `WithCancel` can be done inside `DAG.Get` too.

**Comment by rht** [comment]
`WithTimeout` cleanup should be in a separate PR.

**Comment by rht** [comment]
The remaining cleanup TODO has larger depths, and requires review:
1. p2p/host/basic/natmgr.go#L147: too buried down
2. eventlog: not sure if this needs cleanup
3. merkledag/traverse/traverse.go is not used

Won't touch:
1. test files
2. in exchange/bitswap/testnet/virtual.go `func (n *bitswap.network) SendMessage` because of the comment 'nb: terminate the context since the context wouldn't actually be passed over the network in a real scenario', so won't touch
3. routing/mock/dht.go mockrouting.NewDHTNetwork
4. in routing/supernode/proxy/loopback.go `func (lb *Loopback) HandleStream(s inet.Stream)`, seems to be intentional
5. exchange/bitswap/testnet/peernet.go `func (pn *bitswap.peernet) Adapter` seems to be intentional
6. trickle.VerifyTrickleDagStructure only used in test (and tests already use context.TODO())

File: unixfs/io/dirbuilder.go (hunk 1 above)
Comment (rht) [inline unixfs/io/dirbuilder.go]: remove or replace with `WithCancel`?

**Comment by rht** [comment]
`WithTimeout`-specific question. Should this be moved inside `DAG.Get`? While outside of it, also use `WithCancel`.

**Comment by whyrusleeping** [comment]
1. probably okay to leave for now
2. eventlog doesnt really use contexts for cancellation, so no need to worry about that
3. probably fine not to touch then
4. i think there is a context passed in by the dagbuilderhelper that you might have access to. 

Thanks for doing all this, its really good stuff :)

**Comment by rht** [comment]
4. dagbuilderhelper doesn't have ctx by default. Should it have a ctx field?

**Comment by whyrusleeping** [comment]
@rht hrm.. i'm torn on this one. The only time a context is used is in a situtation where we are expected to already have the node local. We *could* wire a context down through there, but i dont think that its going to be worthwhile (i could be wrong though). What we could do is change the `context.TODO()` to a `context.Background()` and put a comment saying why

**Comment by rht** [comment]
Or just add ctx explicitly as an extra argument, here the ctx flow is:
DagModifier.ctx -> trickle.TrickleAppend -> (n *helpers.UnixfsNode) GetChild

**Comment by rht** [comment]
@whyrusleeping how about this?

Though most of the buried down, long context pipes are constructed because of the need to call dagservice GetNode / GetDAG. It is possibly much cleaner to hide the pipes inside dagservice? (will check if this is the case)

**Comment by rht** [comment]
why it is worthwhile: to make sure there are no 1-min zombie dagservices when an IpfsNode closes.

**Comment by rht** [comment]
> 1. p2p/host/basic/natmgr.go#L147: too buried down

This could be fixed with replacing L120 with `worker.Go(func(p goprocess.Process) {addPortMapping(nmgr, addr)})`

**Comment by rht** [comment]
if err == nil { ready to merge }

**Comment by whyrusleeping** [comment]
LGTM

File: core/core.go [hunk 6]
```diff
@@ -321,9 +321,8 @@ func setupDiscoveryOption(d config.Discovery) DiscoveryOption {
 
 func (n *IpfsNode) HandlePeerFound(p peer.PeerInfo) {
 	log.Warning("trying peer info: ", p)
-	ctx, _ := context.WithTimeout(context.TODO(), time.Second*10)
-	err := n.PeerHost.Connect(ctx, p)
-	if err != nil {
+	ctx, _ := context.WithTimeout(n.Context(), time.Second*10)
```
Comment (jbenet) [inline core/core.go]: oh! found an old bug! we need to always cancel the func.

```go
ctx, cancel := context.WithTimeout(n.Context(), time.Second*10)
defer cancel()
```

Comment (jbenet) [inline core/core.go]: maybe timeout should be `30s` -- high latency links may take much longer to handshake.

Comment (jbenet) [inline core/core.go]: let's make it a constant, too. `discoveryConnTimeout` or something

File: core/coreunix/metadata.go (hunk 4 above)
Comment (jbenet) [inline core/coreunix/metadata.go]: ok makes sense, we can do that.

File: fuse/ipns/common.go [hunk 7]
```diff
@@ -22,7 +20,7 @@ func InitializeKeyspace(n *core.IpfsNode, key ci.PrivKey) error {
 		return err
 	}
 
-	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
+	ctx, cancel := context.WithCancel(n.Context())
```
Comment (jbenet) [inline fuse/ipns/common.go]: maybe use `ctx` in `pub.Publish(ctx, ...)` below in this func.

File: importer/helpers/helpers.go [hunk 8]
```diff
@@ -77,8 +77,8 @@ func (n *UnixfsNode) NumChildren() int {
 	return n.ufmt.NumChildren()
 }
 
-func (n *UnixfsNode) GetChild(i int, ds dag.DAGService) (*UnixfsNode, error) {
-	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
+func (n *UnixfsNode) GetChild(ctx context.Context, i int, ds dag.DAGService) (*UnixfsNode, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
```
Comment (jbenet) [inline importer/helpers/helpers.go]: no timeout, i think?

File: importer/trickle/trickle_test.go [hunk 9]
```diff
@@ -443,7 +443,8 @@ func TestAppend(t *testing.T) {
 	r := bytes.NewReader(should[nbytes/2:])
 	blks, errs := chunk.Chan(chunk.NewSizeSplitter(r, 500))
 
-	nnode, err := TrickleAppend(nd, dbp.New(blks, errs))
+	ctx := context.TODO()
```
Comment (jbenet) [inline importer/trickle/trickle_test.go]: maybe test contexts should be `context.Background()` so we can grep for `context.TODO()`?

**Comment by jbenet** [comment]
@rht some more things to fix
- [ ] cancel func https://github.com/ipfs/go-ipfs/pull/1585/files#r37702176
- [ ] 30s and const https://github.com/ipfs/go-ipfs/pull/1585/files#r37702178
- [ ] move ctx inside dag.Get? https://github.com/ipfs/go-ipfs/pull/1585/files#r37702188
- [ ] ctx in https://github.com/ipfs/go-ipfs/pull/1585/files#diff-17e0497eff558e677d4b3ebb9226b8efR37
- [ ] no timeout? https://github.com/ipfs/go-ipfs/pull/1585/files#r37702194

**Comment by rht** [comment]
>￼ cancel func https://github.com/ipfs/go-ipfs/pull/1585/files#r37702176

This means all lines matching 'ctx, _ *' are to be replaced with 'ctx, cancel ' + `defer cancel()`?

**Comment by rht** [comment]
>￼ move ctx inside dag.Get? https://github.com/ipfs/go-ipfs/pull/1585/files#r37702188

I think this should be done in separate PR. Not every dag.Get caller waits only on this function to finish.

File: fuse/ipns/common.go (hunk 7 above)
Comment (rht) [inline fuse/ipns/common.go]: this hasn't been put in the list, but fixed anyway.

**Comment by rht** [comment]
> >￼ move ctx inside dag.Get? https://github.com/ipfs/go-ipfs/pull/1585/files#r37702188
> I think this should be done in separate PR. Not every dag.Get caller waits only on this function to finish.

Added here anyway. I only made the change here when it is unambiguous that having a `context.WithCancel` before calling dag.Get makes it redundant.

This also "auto-fix" parts of the code where there is no `context.WithCancel`

**Comment by jbenet** [comment]
> This means all lines matching 'ctx, _ *' are to be replaced with 'ctx, cancel ' + defer cancel()?

yeah.

File: namesys/publisher.go [hunk 10]
```diff
@@ -60,7 +60,8 @@ func (p *ipnsPublisher) Publish(ctx context.Context, k ci.PrivKey, value path.Pa
 
 	log.Debugf("Storing pubkey at: %s", namekey)
 	// Store associated public key
-	timectx, _ := context.WithDeadline(ctx, time.Now().Add(time.Second*10))
+	timectx, cancel := context.WithDeadline(ctx, time.Now().Add(time.Second*10))
+	defer cancel()
```
Comment (jbenet) [inline namesys/publisher.go]: - we can probably make this a constant too (and possibly even remove it??) 
- maybe a future PR can go over all the timeouts and make sure they make sense

**Comment by jbenet** [comment]
Ok, this LGTM. the only remaining things:

- [ ] do all commits pass? (maybe run gpe? or squash them if too annoying/slow?)
- [ ] is @whyrusleeping ok with merging this, given dev 0.4.0 and rebase ugliness?

maybe let's make all further refactor/cleanup happen on top of 0.4.0

**Comment by jbenet** [comment]
thanks @rht , this will improve codebase (+ perf for users in slow networkes!) a lot!

File: namesys/publisher.go (hunk 10 above)
Comment (rht) [inline namesys/publisher.go]: Better in another PR. This PR is specifically to make sure all context trees are linked to the root caller.

Comment (jbenet) [inline namesys/publisher.go]: sounds good :+1:

**Comment by rht** [comment]
(+ perf for users in slow networkes!)

> in what way?

**Comment by jbenet** [comment]
requests wont timeout prematurely

**Comment by rht** [comment]
if slow means high latency: lunar RTT is about 2.5s, mars RTT is about 500s-2500s.

**Comment by rht** [comment]
gpe (on rht/go-ipfs) doesn't run the test unless it is done on ipfs/go-ipfs itself.

**Comment by jbenet** [comment]
@rht you have push access to this repo directly. you can push branches here

**Comment by rht** [comment]
1. gpe-ed the commits
2. Just tested a dev0.40 rebase against this branch. There are merge conflicts due to commits for dagmodifier/ipfs add (already in master), but not from this branch.

**Comment by jbenet** [comment]
Ok, sounds good! thanks @rht

**Pull request merged**